### DISCUSSION TITLE
Add a timedelta equivalent for months/years in templates

### DISCUSSION THREAD

**Discussion by koying** [home-assistant/architecture#679] [General]
Discussing on the forums with someone about the difficulties of datetime handling, I came to realize we are really good now but that there is still a gap: months (and maybe years, for consistency) addition / subtractions.

I would happily add that but before coding, it'd be better to have an agreement on the way to do it.
I see 3 options:
- Augment timedelta
- Create datedelta
- Use some 3rd party library

I'm personally inclined towards `datedelta`:
- It keeps timedelta on par with Python
- It's more logical (talking months for *time*delta is a long stretch)
- It's likely less cumbersome to code than overriding classes in Python

Opinions / thoughts?

**Comment by MartinHjelmare** [comment]
Is this about templates?

**Reply by koying** [reply]
Yes, sorry if it wasn't clear.

**Comment by emontnemery** [comment]
I don't think we should implement this ourselves, it's not trivial.
My biggest concern is documenting the behavior for edge cases, for example what is expected when adding a month to January 31st?

There are libraries which allow this, for example `dateutil` which has a `relativedelta`: https://dateutil.readthedocs.io/en/latest/relativedelta.html

**Reply by koying** [reply]
Thanks for the suggestion.

I don't have strong opinions regarding the implementation itself.
What I'd really like is to have some something as clear as possible on the HA template side, i.e. keeping the same kind of  interface/way of working as timedelta.

Time handling is already nightmarish enough without introducing a new interface ;)

**Reply by vlavrynovych** [reply]
Hi, I am not an expert with the Python, but also found that documentation for `relativedelta` and wanted to use it for my `history_stats`  sensors. Wanted to have some easy way to calculate stats for previous month. It's kinda hard to rely on days in calculations knowing how may edge cases we have. `relativedelta` already handles all these cases. Why shouldn't we use it?
